File: backend/src/main/java/com/home/finance_tracker/transaction/entity/Transaction.java
package com.home.finance_tracker.transaction.entity;

import com.home.finance_tracker.category.entity.Category;
import com.home.finance_tracker.subscription.entity.Subscription;
import com.home.finance_tracker.user.entity.User;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;

@Entity
@Table(name = "transactions")
@Getter
@Setter
@NoArgsConstructor
public class Transaction {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(nullable = false, length = 150)
    private String title;

    @Column(nullable = false, precision = 19, scale = 2)
    private BigDecimal amount;

    @Column(nullable = false, length = 3)
    private String currency = "PLN";

    @Column(name = "transaction_date", nullable = false)
    private LocalDate transactionDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "transaction_type", nullable = false, length = 20)
    private TransactionType transactionType;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "subscription_id", nullable = true)
    private Subscription subscription;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;
}

File: frontend/src/types/transaction.ts
import type { CategoryResponseDTO } from './category';

export type TransactionType = 'EXPENSE' | 'INCOME';

export interface TransactionRequestDTO {
  title: string;
  amount: number;
  currency: string;
  transactionDate: string;
  transactionType: TransactionType;
  categoryId: number;
}

export interface TransactionResponseDTO {
  id: number;
  title: string;
  amount: number;
  currency: string;
  transactionDate: string;
  transactionType: TransactionType;
  category: CategoryResponseDTO;
  subscriptionId: number | null;
}